pkg/kubecli/vmi.go

```diff
@@ -245,8 +246,55 @@ func (ws *wsStreamer) Stream(options StreamOptions) error {
 func (v *vmis) VNC(name string) (StreamInterface, error) {
 	return v.asyncSubresourceHelper(name, "vnc")
 }
-func (v *vmis) SerialConsole(name string) (StreamInterface, error) {
-	return v.asyncSubresourceHelper(name, "console")
+
+type connectionStruct struct {
+	con StreamInterface
+	err error
+}
+
+func (v *vmis) SerialConsole(name string, timeout int) (StreamInterface, error) {
+	timeoutChan := time.Tick(time.Duration(timeout) * time.Minute)
+	connectionChan := make(chan connectionStruct)
+
+	go func() {
+		con, err := v.asyncSubresourceHelper(name, "console")
+		for err != nil {
+			if asyncSubresourceError, ok := err.(*AsyncSubresourceError); ok {
+				if asyncSubresourceError.GetStatusCode() == http.StatusBadRequest {
+					// Sleep to prevent denial of service on the api server
+					time.Sleep(500 * time.Millisecond)
+					con, err = v.asyncSubresourceHelper(name, "console")
+				} else {
+					connectionChan <- connectionStruct{con: nil, err: asyncSubresourceError}
+					return
+				}
+			} else {
+				connectionChan <- connectionStruct{con: nil, err: err}
+				return
+			}
+		}
+		connectionChan <- connectionStruct{con: con, err: nil}
+	}()
+
+	select {
+	case <-timeoutChan:
+		return nil, fmt.Errorf("Timeout trying to connect to the virtual machine instance")
```

Comment:
it's nice that we eventually exit but shouldn't we also shut down the goroutine?

Reply:
I don't think we need to stop the goroutine because we have 3 scenarios:
1. the helper return the con or error i will exit the goroutine
2. timeout this return an error an the virctl exit
3. user interrupt the virtctl exit

all of them will stop the goroutine.

Reply:
And when it's used in a functional test, do you terminate the goroutine (before the whole suite exits)? I don't think it's generate a good idea to leak it to outside even if the caller will eventually exit.